apps/els_lsp/src/els_refactorerl_utils.erl

```diff
@@ -0,0 +1,197 @@
+%%  - notconfigured, the node is not configured in the config file
+%%
+%%
+%% Node can be:
+%%  - NodeStr
+%%  - {Status, Node} where both are atoms.
+%%  - Status can be:
+%%    - validated:    node is running
+%%    - disconnected: node is not running
+%%  - disabled: RefactorErl is turned off for this session. T
+%%              his can happen after an unsuccessfull query attempt.
+-spec referl_node() -> {ok, atom()}
+                     | {error, disconnected}
+                     | {error, disabled}
+                     | {error, other}.
+referl_node() ->
+  case els_config:get(refactorerl) of
+    #{"node" := {Node, validated}} ->
+      {ok, Node};
+
+    #{"node" := {Node, disconnected}} ->
+      try_connect_node({retry, Node});
+
+    #{"node" := {_, disabled}} ->
+      {error, disabled};
+
+    #{"node" := NodeStr} ->
+      RT = els_config_runtime:get_name_type(),
+      Node = els_utils:compose_node_name(NodeStr, RT),
+      try_connect_node({validate, Node});
+
+    disabled ->
+      notification("RefactorErl is not configured!"),
+      {error, disabled};
+
+    _ ->
+      {error, other}
+  end.
+
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+-spec disable_node(atom()) -> atom().
+%%@doc
+%% Disables the node given in paramter, also notifes the user.
+disable_node(Node) ->
+  Msg = "RefactorErl is disconnected!
+  Reload ELS after you fixed theRefactorErl node!",
+  notification(Msg, ?MESSAGE_TYPE_ERROR),
+
+  els_config:set(refactorerl, #{"node" => {Node, disabled}}),
+  {error, disabled}.
+
+
+  %%@doc
+  %% Runs a Query on the RefactorErl node, returns its result if successfull
+  %% If error happens, it just returns an empty list, as most of times the
+  %% result then is directly converted to POIs.
+  %% If the node timeouts or badrpc will come back, it disables the node
+-spec query(string()) -> list() | {error, disabled}.
+query(Query) ->
+  case referl_node() of
+    {error, _} ->
+      [];
+    {ok, Node} ->
+      P = [{positions, linecol}, {output, msg}],
+      Response = rpc:call(Node, refusr_sq, run, [P, [], Query], maxtimeout()),
+      case Response of
+        {badrpc, _} ->
+          disable_node(Node); % Returns disabled
+        error ->
+          [];
+        _ ->
+          Response
+      end
+  end.
+
+-spec convert_to_poi(any()) -> poi().
+%%@doc
+%% Convert a RefactorErl result to a POI, for the LS system
+%% The RefactorErl format is how the refusr_sq:run\3 returns
+convert_to_poi(ReferlResult) ->
+case ReferlResult of
+    [{_, _, _, DataList}] ->
+      convert_to_poi(DataList);
+    [{{_, {FromLine, FromCol}, {ToLine, ToCol}}, Name} | Tail] ->
+      Range = #{ from => {FromLine, FromCol}, to => {ToLine, ToCol} },
+      Id = refactorerl_poi, %"{module(), 'atom()', 'arity()''}",
+      Poi = els_poi:new(Range, application, Id, Name),
```

Comment:
Any reason why we generate an intermediate POI here? Why not to generate diagnostics directly from RefactorErl results?

Reply:
Yes I also merged this, it will come soon to this PR ;-)